internal/clients/init.go

```diff
@@ -155,70 +154,74 @@ func checkServiceAvailableViaRegistry(serviceId string) bool {
 	// initialize Core Metadata clients
-	params.ServiceKey = clients.CoreMetaDataServiceKey
-
-	params.Path = clients.ApiAddressableRoute
-	params.Url = metaAddr + params.Path
-	common.AddressableClient = metadata.NewAddressableClient(params, endpoint)
-
-	params.Path = clients.ApiDeviceRoute
-	params.Url = metaAddr + params.Path
-	common.DeviceClient = metadata.NewDeviceClient(params, endpoint)
-
-	params.Path = clients.ApiDeviceServiceRoute
-	params.Url = metaAddr + params.Path
-	common.DeviceServiceClient = metadata.NewDeviceServiceClient(params, endpoint)
-
-	params.Path = clients.ApiDeviceProfileRoute
-	params.Url = metaAddr + params.Path
-	common.DeviceProfileClient = metadata.NewDeviceProfileClient(params, endpoint)
-
-	params.Path = clients.ApiConfigRoute
-	params.Url = metaAddr
-	common.MetadataGeneralClient = general.NewGeneralClient(params, endpoint)
-
-	params.Path = clients.ApiProvisionWatcherRoute
-	params.Url = metaAddr + params.Path
-	common.ProvisionWatcherClient = metadata.NewProvisionWatcherClient(params, endpoint)
+	common.AddressableClient = metadata.NewAddressableClient(
+		urlclient.NewMetadata(
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiAddressableRoute,
+		),
+	)
+
+	common.DeviceClient = metadata.NewDeviceClient(
+		urlclient.NewMetadata(
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiDeviceRoute,
+		),
+	)
+
+	common.DeviceServiceClient = metadata.NewDeviceServiceClient(
+		urlclient.NewMetadata(
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiDeviceServiceRoute,
+		),
+	)
+
+	common.DeviceProfileClient = metadata.NewDeviceProfileClient(
+		urlclient.NewMetadata(
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiDeviceProfileRoute,
+		),
+	)
+
+	common.MetadataGeneralClient = general.NewGeneralClient(
+		urlclient.NewMetadata(
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiConfigRoute,
+		),
+	)
+
+	common.ProvisionWatcherClient = metadata.NewProvisionWatcherClient(
+		urlclient.NewMetadata(
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiProvisionWatcherRoute,
+		),
+	)
 
 	// initialize Core Data clients
-	params.ServiceKey = clients.CoreDataServiceKey
-
-	params.Path = clients.ApiEventRoute
-	params.Url = dataAddr + params.Path
-	common.EventClient = coredata.NewEventClient(params, endpoint)
-
-	params.Path = common.APIValueDescriptorRoute
-	params.Url = dataAddr + params.Path
-	common.ValueDescriptorClient = coredata.NewValueDescriptorClient(params, endpoint)
-
-	if isRegistry {
+	common.EventClient = coredata.NewEventClient(
+		urlclient.NewData(
+			common.RegistryClient,
+			waitGroup,
+			clients.ApiEventRoute,
+		),
+	)
+
+	common.ValueDescriptorClient = coredata.NewValueDescriptorClient(
+		urlclient.NewData(
+			common.RegistryClient,
+			waitGroup,
+			common.APIValueDescriptorRoute,
+		),
+	)
+
+	if common.RegistryClient != nil {
 		// wait for the first endpoint discovery to make sure all clients work
 		waitGroup.Wait()
```

Comment:
Not sure if we can use the global waitGroup here. I tried to ran device-simple with `-r` argument and it'll block here because there're other long-running goroutines utilize this waitGroup in bootstrap package ([for example](https://github.com/edgexfoundry/go-mod-bootstrap/blob/3ab2457cb6c9e600d5c81a556494d09cad27121c/bootstrap/bootstrap.go#L49))

Reply:
We need the global wait group to signal to the URL Client gofuncs that it's time to end; should this be a second wait group or do we even need to wait here?

Reply:
I think we should use a second wait group here per the comment suggests(wait for the first endpoint discovery), @lenny-intel any thought about this ?

Reply:
Where is this wait group used though? If we're just waiting for all the clients to resolve, we don't need a wait group for that. The URL client handles that for us.

Reply:
> Where is this wait group used though? If we're just waiting for all the clients to resolve, we don't need a wait group for that. The URL client handles that for us.

In this case I think we should not call `waitGroup.wait()` here